docker/apm-server/Dockerfile

```diff
@@ -7,14 +7,24 @@ FROM golang:${go_version} AS build
 RUN apt-get -qq update && apt-get -qq install -y python-virtualenv
 
 ARG apm_server_branch=master
+ARG apm_server_commit=
 ARG apm_server_repo=https://github.com/elastic/apm-server.git
-RUN git clone -b $apm_server_branch ${apm_server_repo} /go/src/github.com/elastic/apm-server && \
-	make -C /go/src/github.com/elastic/apm-server update apm-server && \
-	sed -zri -e 's/output.elasticsearch:(\n[^\n]*){5}/output.elasticsearch:\n  hosts: ["elasticsearch:9200"]/' -e 's/  host: "localhost:8200"/  host: "0.0.0.0:8200"/' /go/src/github.com/elastic/apm-server/apm-server.yml && \
-	chmod go+r /go/src/github.com/elastic/apm-server/apm-server.yml
+ENV SRC=/go/src/github.com/elastic/apm-server
+
+# Git clone and checkout given either the branch, commit or both.
+RUN [ -z "${apm_server_commit}" ] \
```

Comment:
you can use the commit approach also with branches 

```
git clone ${apm_server_repo} ${SRC}
cd ${SRC} 
git reset --hard ${apm_server_branch_or_commit} // or git checkout -b ${apm_server_branch_or_commit} origin/${apm_server_branch_or_commit}

Reply:
When using a branch then the git reset --hard doesn't works as expected though
```
Step 8/14 : RUN git clone ${apm_server_repo} ${SRC}     && cd ${SRC} && git reset --hard ${apm_server_branch_or_commit}
 ---> Running in 77b42e2e4796
Cloning into '/go/src/github.com/elastic/apm-server'...
fatal: ambiguous argument 'feature/triggerITsAsync': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

I tried locally though:

```
root@edda09baa2e6:/go/src/github.com/elastic/apm-server# git remote -v
origin	https://github.com/v1v/apm-server (fetch)
origin	https://github.com/v1v/apm-server (push)
root@edda09baa2e6:/go/src/github.com/elastic/apm-server# git reset --hard ${apm_server_branch_or_commit}
fatal: ambiguous argument 'feature/triggerITsAsync': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

Reply:
I cannot find when to distinguish whether `origin` should be in place when doing either the reset or the checkout -b

Reply:
I've forgotten the fetch, see https://github.com/elastic/apm-integration-testing/blob/master/docker/java/spring/Dockerfile#L15-L17